refactor: adopt some hexagonal architecture components to better organize the core package

d-klotz opened the pull request:
This PR introduces some DDD and Hexagonal Architecture components to better organize Frigg Core.
Currently there are a lot of unmapped dependencies between the core files, database calls happening in all places and a lot of multiple responsibilities for single methods.

Our goal is to make use of the Use Case, Repository and IoC concepts and possibly adopting Ports and Adapters. This should give us a more organized code, making clear which responsibility belongs to which component and creating better boundaries to data flow.

https://link.excalidraw.com/l/4ihrHVE4BFS/9hLNCzJMCbt

<!-- GITHUB_RELEASE PR BODY: canary-version -->
<details>
  <summary>📦 Published PR as canary version: <code>2.0.0--canary.395.a8332aa.0</code></summary>
  <br />

  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install @friggframework/core@2.0.0--canary.395.a8332aa.0
  npm install @friggframework/devtools@2.0.0--canary.395.a8332aa.0
  npm install @friggframework/eslint-config@2.0.0--canary.395.a8332aa.0
  npm install @friggframework/prettier-config@2.0.0--canary.395.a8332aa.0
  npm install @friggframework/schemas@2.0.0--canary.395.a8332aa.0
  npm install @friggframework/serverless-plugin@2.0.0--canary.395.a8332aa.0
  npm install @friggframework/test@2.0.0--canary.395.a8332aa.0
  npm install @friggframework/ui@2.0.0--canary.395.a8332aa.0
  # or 
  yarn add @friggframework/core@2.0.0--canary.395.a8332aa.0
  yarn add @friggframework/devtools@2.0.0--canary.395.a8332aa.0
  yarn add @friggframework/eslint-config@2.0.0--canary.395.a8332aa.0
  yarn add @friggframework/prettier-config@2.0.0--canary.395.a8332aa.0
  yarn add @friggframework/schemas@2.0.0--canary.395.a8332aa.0
  yarn add @friggframework/serverless-plugin@2.0.0--canary.395.a8332aa.0
  yarn add @friggframework/test@2.0.0--canary.395.a8332aa.0
  yarn add @friggframework/ui@2.0.0--canary.395.a8332aa.0
  ```
</details>
<!-- GITHUB_RELEASE PR BODY: canary-version -->
<!-- GITHUB_RELEASE PR BODY: prerelease-version -->
# Version

Published prerelease version: `v2.0.0-next.42`

<details>
  <summary>Changelog</summary>

  #### 🚀 Enhancement
  
  - `@friggframework/core`, `@friggframework/devtools`, `@friggframework/schemas`, `@friggframework/ui`
    - refactor: adopt some hexagonal architecture components to better organize the core package [#395](https://github.com/friggframework/frigg/pull/395) ([@d-klotz](https://github.com/d-klotz))
  - `@friggframework/core`, `@friggframework/devtools`, `@friggframework/eslint-config`, `@friggframework/prettier-config`, `@friggframework/schemas`, `@friggframework/serverless-plugin`, `@friggframework/test`, `@friggframework/ui`
    - Skip AWS discovery when running Frigg start locally [#431](https://github.com/friggframework/frigg/pull/431) ([@seanspeaks](https://github.com/seanspeaks))
    - feat: add KMS decrypt capability check to health endpoint [#428](https://github.com/friggframework/frigg/pull/428) ([@d-klotz](https://github.com/d-klotz) [@seanspeaks](https://github.com/seanspeaks))
  - `@friggframework/core`, `@friggframework/devtools`, `@friggframework/eslint-config`, `@friggframework/prettier-config`, `@friggframework/serverless-plugin`, `@friggframework/test`, `@friggframework/ui`
    - Feat/improve-infrastructure [#405](https://github.com/friggframework/frigg/pull/405) ([@d-klotz](https://github.com/d-klotz))
    - feat: add KMS support for frigg framework [#392](https://github.com/friggframework/frigg/pull/392) ([@d-klotz](https://github.com/d-klotz))
  
  #### 🐛 Bug Fix
  
  - `@friggframework/core`, `@friggframework/devtools`, `@friggframework/eslint-config`, `@friggframework/prettier-config`, `@friggframework/schemas`, `@friggframework/serverless-plugin`, `@friggframework/test`, `@friggframework/ui`
    - fix: add API Gateway v2 permissions for HttpApi creation [#410](https://github.com/friggframework/frigg/pull/410) ([@seanspeaks](https://github.com/seanspeaks))
  - `@friggframework/core`, `@friggframework/devtools`, `@friggframework/eslint-config`, `@friggframework/prettier-config`, `@friggframework/serverless-plugin`, `@friggframework/test`, `@friggframework/ui`
    - refactor: update serverless template to use HTTP API and streamline CORS configuration [#406](https://github.com/friggframework/frigg/pull/406) ([@d-klotz](https://github.com/d-klotz))
  - `@friggframework/core`, `@friggframework/devtools`, `@friggframework/ui`
    - turn module factory in a set of use cases [#397](https://github.com/friggframework/frigg/pull/397) ([@d-klotz](https://github.com/d-klotz))
    - refactor: turn integration factory in a set of use cases. [#396](https://github.com/friggframework/frigg/pull/396) ([@d-klotz](https://github.com/d-klotz))
  
  #### Authors: 2
  
  - Daniel Klotz ([@d-klotz](https://github.com/d-klotz))
  - Sean Matthews ([@seanspeaks](https://github.com/seanspeaks))
</details>
<!-- GITHUB_RELEASE PR BODY: prerelease-version -->

d-klotz committed:
refactor: remove createFriggBackend and related middleware

- Removed `createFriggBackend` from core integrations and updated related files to eliminate its usage.
- Refactored app-handler-helpers to remove the `loadUserManager` middleware.
- Introduced `loadAppDefinition` function in backend-utils to load application definitions from the nearest backend package.
- Updated routers to utilize the new `loadAppDefinition` function for integration and user management.
- Cleaned up unused middleware files and adjusted user handling in various routers.

seanspeaks commented on packages/core/handlers/backend-utils.js:
```diff
@@ -1,24 +1,40 @@
+/**
+ * Loads the App definition from the nearest backend package
+ * @function loadAppDefinition
+ * @description Searches for the nearest backend package.json, loads the corresponding index.js file,
```
As we restructure by removing the "frontend" directory from the default frigg structure, this makes me realize we need to rethink defaults and likely provide an override somewhere to point at the frigg entry point/definition. Frigg config comes to mind, lots of JS libraries and tools do that ("use the Frigg config file to tell us where to look/find things").

seanspeaks commented on packages/core/handlers/routers/auth.js:
```diff
@@ -1,26 +1,36 @@
+    getUserFromBearerToken,
 });
 
+//todo: what is this route doing here?
```
Backend redirect cases, I think we introduced via the Canva use case where we needed to redirect people to our service that we controlled instead of to the frontend of the core application. That happens  lot in apps like Asana, Google workspace, etc. 

Our more recent solutions have been to do that redirect handling in the integration defined router, which didn't exist as a concept at the time of building this thing out.

seanspeaks commented on packages/core/handlers/routers/auth.js:
```diff
@@ -1,26 +1,36 @@
 const handler = createAppHandler('HTTP Event: Auth', router);
 
+// todo: I can not find where router is used, do we need to export it?
 module.exports = { handler, router };
```
Maybe check the canva repo?

I can't recall what prompted keeping this as an export.

d-klotz replied:
@seanspeaks My question was more on the "shouldn't this live somewhere else ?" direction.

seanspeaks commented on packages/core/handlers/routers/user.js:
```diff
@@ -1,20 +1,39 @@
 
 const router = express();
 
+// A poor-man's dependency injection container
```
Ah I see. Hadn't seen this pattern before. Downsides? More verbose/scaffolding in a given file?

seanspeaks reviewed (commented):
Three some comments in. Pausing for now.

d-klotz committed:
feat: improve user class and add more tests

d-klotz committed:
feat: improve jsdocs

d-klotz committed:
feat: remove unused methods

d-klotz committed:
feat: iterate on User class

d-klotz committed:
test: fix tests

d-klotz committed:
test: improve test

d-klotz committed:
refactor: remove userFactory

d-klotz committed:
refactor:  userRepository

d-klotz committed:
feat: add missing dependency

d-klotz committed:
fix: add missing parameter

d-klotz committed:
fix: remove param

d-klotz committed:
refactor: move user instantiation to the use cases

d-klotz committed:
refactor: remove comment

d-klotz committed:
feat: remove coments

d-klotz committed:
fix: forgot to re-add import

d-klotz committed:
fix: pass correct params

d-klotz committed:
fix: remove unnecessary param

d-klotz committed:
refactor: turn integration factory in a set of use cases.

d-klotz committed:
turn module factory in a set of use cases

d-klotz committed:
refactor: turn integration factory in a set of use cases. (#396)

<!-- GITHUB_RELEASE PR BODY: canary-version -->
<details>
  <summary>📦 Published PR as canary version: <code>2.0.0--canary.396.469364a.0</code></summary>
  <br />

  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install @friggframework/core@2.0.0--canary.396.469364a.0
  npm install @friggframework/devtools@2.0.0--canary.396.469364a.0
  npm install @friggframework/eslint-config@2.0.0--canary.396.469364a.0
  npm install @friggframework/prettier-config@2.0.0--canary.396.469364a.0
  npm install @friggframework/serverless-plugin@2.0.0--canary.396.469364a.0
  npm install @friggframework/test@2.0.0--canary.396.469364a.0
  npm install @friggframework/ui@2.0.0--canary.396.469364a.0
  # or 
  yarn add @friggframework/core@2.0.0--canary.396.469364a.0
  yarn add @friggframework/devtools@2.0.0--canary.396.469364a.0
  yarn add @friggframework/eslint-config@2.0.0--canary.396.469364a.0
  yarn add @friggframework/prettier-config@2.0.0--canary.396.469364a.0
  yarn add @friggframework/serverless-plugin@2.0.0--canary.396.469364a.0
  yarn add @friggframework/test@2.0.0--canary.396.469364a.0
  yarn add @friggframework/ui@2.0.0--canary.396.469364a.0
  ```
</details>
<!-- GITHUB_RELEASE PR BODY: canary-version -->

d-klotz committed:
turn module factory in a set of use cases (#397)

<!-- GITHUB_RELEASE PR BODY: canary-version -->
<details>
  <summary>📦 Published PR as canary version: <code>2.0.0--canary.397.c07f148.0</code></summary>
  <br />

  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install @friggframework/core@2.0.0--canary.397.c07f148.0
  npm install @friggframework/devtools@2.0.0--canary.397.c07f148.0
  npm install @friggframework/eslint-config@2.0.0--canary.397.c07f148.0
  npm install @friggframework/prettier-config@2.0.0--canary.397.c07f148.0
  npm install @friggframework/serverless-plugin@2.0.0--canary.397.c07f148.0
  npm install @friggframework/test@2.0.0--canary.397.c07f148.0
  npm install @friggframework/ui@2.0.0--canary.397.c07f148.0
  # or 
  yarn add @friggframework/core@2.0.0--canary.397.c07f148.0
  yarn add @friggframework/devtools@2.0.0--canary.397.c07f148.0
  yarn add @friggframework/eslint-config@2.0.0--canary.397.c07f148.0
  yarn add @friggframework/prettier-config@2.0.0--canary.397.c07f148.0
  yarn add @friggframework/serverless-plugin@2.0.0--canary.397.c07f148.0
  yarn add @friggframework/test@2.0.0--canary.397.c07f148.0
  yarn add @friggframework/ui@2.0.0--canary.397.c07f148.0
  ```
</details>
<!-- GITHUB_RELEASE PR BODY: canary-version -->

d-klotz committed:
Merge branch 'next' into feat/general-code-improvements

d-klotz committed:
Merge branch 'next' into feat/general-code-improvements

d-klotz committed:
Merge branch 'next' into feat/general-code-improvements

d-klotz committed:
Merge branch 'next' into feat/general-code-improvements

d-klotz committed:
refactor: enhance shouldRunDiscovery function to skip AWS discovery in offline mode

d-klotz committed:
Revert "refactor: enhance shouldRunDiscovery function to skip AWS discovery in offline mode"

This reverts commit eb0264e7fb4d26ee5fa81fb693905775609e5671.

d-klotz committed:
Merge branch 'next' into feat/general-code-improvements

d-klotz committed:
feat: implement IntegrationEventDispatcher for handling integration events and refactor related classes

d-klotz committed:
feat: Enhance integration context loading and module hydration

- Added LoadIntegrationContextUseCase to manage loading integration contexts with associated modules.
- Implemented setIntegrationRecord method in IntegrationBase for better hydration of integration instances.
- Created comprehensive tests for LoadIntegrationContextUseCase to ensure proper loading of integration records and module instances.
- Introduced ModuleFactory and ModuleRepository for managing module instances and definitions.
- Added support for handling missing credentials gracefully in module instances.
- Enhanced module hydration tests to verify API instantiation with environment variables and credentials.
- Implemented findEntitiesByUserIdAndModuleName method in ModuleRepository for querying entities by user ID and module name.

d-klotz committed:
fix: Correct import path for Credential module in CredentialRepository

d-klotz committed:
fix: disable mongodb strict schema for credentials

seanspeaks commented on packages/core/core/create-handler.js:
```diff
@@ -34,6 +34,7 @@ const createHandler = (optionByName = {}) => {
             // Helps mongoose reuse the connection.  Lowers response times.
             context.callbackWaitsForEmptyEventLoop = false;
 
+            // todo: this should not be necessary anymore
```
Kk. There are cases where we definitely don't want to, which I believe falls under.... a use case? an adapter?

d-klotz replied:
Yes, correct. Probably the right place to decide if we want to connect to a database or not is inside of the specific use case for it.

seanspeaks commented on packages/core/credential/use-cases/get-credential-for-user.js:
```diff
@@ -0,0 +1,21 @@
+class GetCredentialForUser {
```
Just a curiosity question, do all use cases get Classes?

d-klotz replied:
I just pasted the answer from ChatGPT here :
```
no. In hexagonal architecture a “use case” is a boundary (an application-level operation) — not inherently a class. You can model it as a class, a function, a module, or a command-handler, as long as:
	•	it exposes a clear input/output shape (often simple DTOs),
	•	orchestrates the domain and calls out only through ports,
	•	is easy to unit test in isolation (TDD-friendly).

```

In my personal opinion, using classes gets more powerful when you're using TypeScript for example. Then you can actually say that a use case depends on an interface and not straight to a class like we have to do here in JavaScript for example. And with that, we can basically reuse a use case and pass it in different implementations of what the constructor needs. And with that, testing is also easier since we can basically create our own testing versions of what the constructor needs.

seanspeaks replied:
When do we do TS migration, is my question 😂  I've been appreciating having the "schemas" repo that I can point LLMs at, and running ts checking saves a whole lot of heartache/headache (though making sure it doesn't just :any everywhere and also doesn't make up the objects/types becomes the critical path)

seanspeaks commented on packages/core/credential/credential-repository.js:
```diff
@@ -0,0 +1,56 @@
+            strict: false,
+        };
+
+        const credential = await Credential.findOneAndUpdate(query, update, options);
```
This gets updated with the adapters, yeah?

d-klotz replied:
Yes, correct. So here probably the credential repository will use just the Prisma ORM methods instead of accessing the DB straight away.

seanspeaks commented on packages/core/handlers/routers/user.js:
```diff
@@ -13,8 +32,8 @@ router.route('/user/login').post(
         ]);
-        const user = await User.loginUser({ username, password });
-        const token = await user.createUserToken(120);
+        const user = await loginUser.execute({ username, password });
```
Apologies this is just me making sure I understand the hexagonal basics, but... .execute is basically the only method we'd call on every use case class, yeah? When do other methods get called?

d-klotz replied:
Yes, the public surface should be one operation—usually execute/handle/run. Everything else is private helpers. If you find yourself wanting more public methods, that’s a good sign you actually have multiple use cases.

d-klotz replied:
Also falls under SRP

seanspeaks commented on packages/core/handlers/auth-flow.integration.test.js:
```diff
@@ -0,0 +1,140 @@
+    constructor(params = {}) {
+        super(params);
+        this.events = {
+            AUTH_REQUEST: { handler: this.authRequest.bind(this) },
```
We should throw in user actions at some point but I'll defer for the moment

seanspeaks commented on packages/core/handlers/auth-flow.integration.test.js:
```diff
@@ -0,0 +1,140 @@
+
+describe('IntegrationEventDispatcher auth flow', () => {
+    const createDispatcher = () =>
+        new IntegrationEventDispatcher(new SimulatedAsanaIntegration());
```
"What is an event dispatcher?" is an answer I will get in 2 files 😂

d-klotz replied:
So, before we had the backend, we used that to import the integration and check what available routes (integration routes) were there. Then, given the actual request path, we were invoking the method that matches that specific route. And that logic was a bit confusing. So what I did is I created an integration event dispatcher that should make it clear what it does (following SRP). It receives the route and it receives integration and we just dispatch that event.

seanspeaks commented on packages/core/handlers/integration-event-dispatcher.js:
```diff
@@ -0,0 +1,55 @@
+      throw new Error(`Event ${event} not registered for ${name}`);
+    }
+
+    return await handler.call(instance, { data, context });
```
Probably a linting thing at a later date but return await in an async function is unnecessary, so says my linters that bark at me

seanspeaks commented on packages/core/integrations/tests/doubles/dummy-integration-class.js:
```diff
@@ -0,0 +1,90 @@
+    };
+}
+
+class DummyIntegration extends IntegrationBase {
```
What's a test double?

d-klotz replied:
A test double is a stand-in object you use in a test in place of a real collaborator. It lets you isolate the system under test, control hard-to-reproduce situations, and make tests fast and deterministic (e.g., no real network, clock, DB).

seanspeaks reviewed (commented):
This is starting to make sense to me and feels directionally correct.

d-klotz committed:
feat: Implement integration and user command factories

- Added integration commands for managing integration contexts, including loading and finding by external entity ID.
- Created user commands for CRUD operations on user entities.
- Introduced a unified command factory to streamline access to integration, user, entity, and credential commands.
- Enhanced integration event dispatcher for better event handling.
- Updated credential repository with new methods for finding and updating credentials.
- Added tests for new integration use cases and command functionalities.
- Refactored existing code for improved readability and maintainability.

d-klotz committed:
Refactor WebSocket and Integration Handling

- Updated WebSocket connection handling to use WebsocketConnectionRepository for creating and deleting connections.
- Refactored IntegrationBase to utilize IntegrationMappingRepository for mapping operations.
- Introduced IntegrationMappingRepository to encapsulate integration mapping logic.
- Enhanced SyncManager to leverage SyncRepository for sync operations, improving code organization and readability.
- Created SyncRepository to manage sync object persistence and operations.
- Refactored UserRepository to utilize TokenRepository for token management, enhancing modularity.

d-klotz committed:
feat: Implement Prisma-based User Repository

- Added UserRepository class implementing UserRepositoryInterface for user operations.
- Migrated user creation and retrieval methods from Mongoose to Prisma.
- Introduced TokenRepository for session token management.
- Updated use cases to reference UserRepositoryInterface instead of the deprecated UserRepository.
- Removed legacy UserRepository implementation.
- Added Websocket Connection Repository and its interface for managing WebSocket connections.
- Created infrastructure documentation for Frigg Framework, detailing architecture, deployment, and security practices.

d-klotz committed:
feat: Implement PostgreSQL and MongoDB repository

- Added TokenRepositoryPostgres for handling token persistence with PostgreSQL.
- Created UserRepositoryMongo and UserRepositoryPostgres for user operations with MongoDB and PostgreSQL respectively.
- Introduced WebsocketConnectionRepositoryMongo and WebsocketConnectionRepositoryPostgres for managing WebSocket connections in both databases.
- Updated user-repository-factory and websocket-connection-repository-factory to support new database-specific implementations.
- Ensured all repository methods return string IDs for consistency across application layers.

seanspeaks commented:
* **#452** <a href="https://app.graphite.dev/github/pr/friggframework/frigg/452?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#450** <a href="https://app.graphite.dev/github/pr/friggframework/frigg/450?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#449** <a href="https://app.graphite.dev/github/pr/friggframework/frigg/449?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#446** <a href="https://app.graphite.dev/github/pr/friggframework/frigg/446?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#445** <a href="https://app.graphite.dev/github/pr/friggframework/frigg/445?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#444** <a href="https://app.graphite.dev/github/pr/friggframework/frigg/444?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#443** <a href="https://app.graphite.dev/github/pr/friggframework/frigg/443?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#442** <a href="https://app.graphite.dev/github/pr/friggframework/frigg/442?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#441** <a href="https://app.graphite.dev/github/pr/friggframework/frigg/441?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#440** <a href="https://app.graphite.dev/github/pr/friggframework/frigg/440?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#439** <a href="https://app.graphite.dev/github/pr/friggframework/frigg/439?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#438** <a href="https://app.graphite.dev/github/pr/friggframework/frigg/438?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#437** <a href="https://app.graphite.dev/github/pr/friggframework/frigg/437?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#436** <a href="https://app.graphite.dev/github/pr/friggframework/frigg/436?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#395** <a href="https://app.graphite.dev/github/pr/friggframework/frigg/395?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/friggframework/frigg/395?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `next`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

d-klotz committed:
refactor: Simplify credential ID handling in PostgreSQL repository and module

d-klotz committed:
Refactor encryption module and health check routes

- Removed the Encrypt class from the encryption module, retaining only the Cryptor class.
- Updated test-encrypt.js to remove references to the Encrypt class.
- Refactored health.js to integrate new use cases for checking encryption and external APIs health.
- Created CheckExternalApisHealthUseCase and CheckIntegrationsHealthUseCase to encapsulate health check logic.
- Removed unused external API and integration checks from health.js.
- Updated health check routes to utilize the new use case classes for better separation of concerns.
- Deleted the Credential model as it was dependent on the removed Encrypt class.
- Cleaned up index.js and module index files to remove references to the deleted Credential model.

seanspeaks commented on packages/core/database/encryption/README.md:
```diff
@@ -0,0 +1,496 @@
+- ✅ **Transparent**: Repositories and use cases work with plain data
+- ✅ **Hexagonal architecture**: Clean separation of concerns
+- ✅ **AWS KMS support**: Enterprise-grade encryption with AWS Key Management Service
+- ✅ **Local AES fallback**: Development mode using local encryption keys
```
Not quite "development mode". We should skip encyrption for development. AES_KEY can work in production, it's a no-cost option that requires manual key rotation. (TBD on if that's preserved, looking now...)

seanspeaks commented on packages/core/database/encryption/README.md:
```diff
@@ -0,0 +1,496 @@
+- ✅ **Hexagonal architecture**: Clean separation of concerns
+- ✅ **AWS KMS support**: Enterprise-grade encryption with AWS Key Management Service
+- ✅ **Local AES fallback**: Development mode using local encryption keys
+- ✅ **Environment-based**: Automatic bypass in dev/test/local environments
```
👌

seanspeaks commented on packages/core/database/encryption/README.md:
````diff
@@ -0,0 +1,496 @@
+#### Development (Local AES)
+
+```bash
+# Local AES encryption (development/testing only)
````
Yeah this needs to be updated, misnomer

seanspeaks commented on packages/core/database/encryption/README.md:
````diff
@@ -0,0 +1,496 @@
+STAGE=development  # or dev, test, local
+```
+
+Or simply don't configure any encryption keys.
````
This should error in Prod, I think? Breaks and doesn't work

seanspeaks commented on packages/core/database/encryption/README.md:
````diff
@@ -0,0 +1,496 @@
+
+```javascript
+// 1. Define custom Prisma model (in your backend prisma schema)
+model AsanaTaskMapping {
````
Would we do this or just use integrationMapping?

d-klotz replied:
This is an example of a custom Prisma model.

seanspeaks commented on packages/core/database/encryption/README.md:
```diff
@@ -0,0 +1,496 @@
+
+### Extending Encryption Schema
+
+#### Recommended: Custom Schema via appDefinition (Integration Developers)
```
Prompt Claude to review friggframework/api-module-library (v1-ready) or the api module definition Schema and determine where in the Definition we can allow api modules to specify whcih of their fields to encrypt.

I'm okay with appDefinition level extensions, but each API module should also be able to indicate which fields it wants encrypted via the same method and the framework core should just aggregrate those automatically.

d-klotz replied:
Added a task for that: https://app.clickup.com/t/86acbc0j9

seanspeaks commented on packages/core/database/encryption/README.md:
````diff
@@ -0,0 +1,496 @@
+```javascript
+// Application code (use case or repository)
+await prisma.credential.create({
+    data: {
````
Why is it data data?

d-klotz replied:
first "data" is just Prisma API. the second "data" is our field, in this example.
But It could also be `data.userId`.

d-klotz committed:
Refactor repository factories and implementations to remove optional Prisma client injection

- Updated repository factories (e.g., User, Integration, Sync, Token, Websocket) to remove optional Prisma client parameter, using default Prisma instance instead.
- Adjusted repository constructors to eliminate unnecessary parameters, simplifying instantiation.
- Mocked database configuration in tests to ensure consistent behavior across different test files.
- Removed unused user repository file and related tests to streamline the codebase.
- Enhanced error messages for unsupported database types for better clarity.

d-klotz committed:
refactor: Update credential structure to encapsulate tokens within a data object

d-klotz committed:
feat: integrate prisma to Frigg CLI

- Added error messages utility for better user feedback during database setup.
- Created database validator functions to validate DATABASE_URL and check database type.
- Implemented Prisma runner functions to handle Prisma CLI commands for generating clients and running migrations.
- Developed a mock Prisma client for testing purposes.
- Created unit tests for error messages and database validation utilities.
- Integrated dotenv for loading environment variables from .env file.
- Enhanced user experience with detailed troubleshooting steps for database connection issues.

d-klotz committed:
feat(tests): add comprehensive unit tests for database validation and Prisma runner utilities

- Implement unit tests for database-validator including validation of DATABASE_URL, database type retrieval, and connection testing.
- Add tests for Prisma runner utilities covering schema path retrieval, Prisma generate execution, database state checking, migration commands, and database push operations.
- Mock dependencies for child processes and file system interactions to ensure isolated test environments.
- Enhance error handling in tests to cover various failure scenarios and ensure proper logging.
- Update README and schema definitions to maintain consistency in naming conventions.

d-klotz committed:
feat: use Prisma database name convention

d-klotz committed:
Refactor IAM generator and tests; enhance CLI command tests

- Refactored IAM CloudFormation template generation to accept flattened options and utilize feature summary extraction.
- Updated tests for IAM generator to reflect new API and ensure accurate YAML structure.
- Removed unused CommandTester utility class to streamline test setup.
- Improved test setup for CLI commands by mocking necessary dependencies and capturing console output.
- Added Jest configuration for better test coverage reporting and management.
- Enhanced start command tests to prevent actual process exits and ensure proper error handling.
- Updated generate command tests to validate actual template generation for AWS services.
- Cleaned up console warning suppression in test setup to avoid noise during test execution.

d-klotz committed:
refactor: update Prisma schema paths and enhance database validation checks

d-klotz committed:
feat(tests): enhance db setup command tests for Prisma client generation logic

d-klotz committed:
Add tests for PostgreSQL decryption fixes and enhance repository methods

- Introduced `postgres-decryption-fix-verification.test.js` to verify that the fix in `ModuleRepositoryPostgres` correctly decrypts credentials when fetching entities.
- Added `postgres-relation-decryption.test.js` to demonstrate the bug where credentials fetched via Prisma `include` relations are not decrypted.
- Enhanced `ModuleRepositoryMongo` and `ModuleRepositoryPostgres` to fetch credentials separately, ensuring decryption occurs correctly.
- Updated `frigg-cli` to skip database connection testing during the start command to avoid module resolution issues.
- Removed database connection tests from `start-command.test.js` to reflect the changes in the CLI behavior.

d-klotz committed:
docs: update encryption documentation to clarify AWS KMS and AES usage

d-klotz committed:
feat: enhance field encryption and decryption to handle object serialization and deserialization

d-klotz committed:
feat: add comprehensive documentation for Frigg Framework core package

seanspeaks commented:
<!-- GITHUB_RELEASE COMMENT: released -->
:rocket: PR was released in [`v2.0.0-next.42`](https://github.com/friggframework/frigg/releases/tag/v2.0.0-next.42) :rocket:

seanspeaks commented:
<!-- GITHUB_RELEASE COMMENT: released -->
:rocket: PR was released in [`v2.0.0-next.46`](https://github.com/friggframework/frigg/releases/tag/v2.0.0-next.46) :rocket:
